Rebuilding based on upstream

aaronbell opened the pull request:
Hello!

I've been working to 'rebase' the build for Chocolate Sans to be based on your upstream repository following the same methodology I implemented for [UoqMunThenKhung](https://github.com/MoonlitOwen/ThenKhung/pull/6) and wanted to provide you with a PR of the changes I made, in case you'd like to merge them back upstream, or have any questions / comments.

*Note, all source changes have been made in the .fcp source with version 15.0.0.2978. Hopefully this means it'll be easy for you to work on in the future.*

- added build system following ThenKhung model
- Added BASE table
- Added copyright info to license string
- Added AUTHORs file
- un-map softhyphen
- vertical metrics updated to follow GF recommendation (If you have a preference for these to be set differently, let's talk! I'm happy to discuss.)
- version string metadata removed from source. Now done in the config file. 
- update postscript name per GF requirements
- renamed some very long glyph names:
* katakanasmallhu_katakanahiraganasemivoicedsoundmarkcombkana.ccmp
* katakanasmallhu_katakanahiraganasemivoicedsoundmarkcombkana.ccmp.1 

- Adjust anchor in glyphs and add anchors to dotted circle (u25CC)
uni0300                                                                                                                                                               
uni0301                                                                                                                                                               
uni0307                                                                                                                                                               
uni030C

- uni9E44 - fixed overlap issue due to overlapping contours where is quadratic and one is cubic. 

- Cleared overlapping segments / duplicate contours on various glyphs
uni50C6
uni5966
uni79BB
uni7E34
uni8506
uni8867
uni9C40

---

**FB FAILS**
- file_size
TC fonts are big! :)

- googlefonts/glyphsets/shape_languages
Font not intended for Vietnamese support. 

- googlefonts/family_name_compliance
Localized name. Not a problem.

- googlefonts/cjk_vertical_metrics_regressions
Check has not been updated with new CJK metrics requirements

- googlefonts/glyph_coverage
Font contains Kernal coverage



---
Please let me know if you have any questions, or if there's anything I can help explain.

aaronbell commented:
Hi. Just following up to see if you have any thoughts about the font. Any changes needed?

NightFurySL2001 commented:
I need to report back a current bug in FC 15.0.0.2978: the `public.unicodeVariationSequences` key exported in the UFO file by FC doesn't use the correct format:

<details>
  <summary>Incorrect `public.unicodeVariationSequences` format exported by FCFC 15.0.0.2978</summary>

```xml
<key>public.unicodeVariationSequences</key>
  <dict>
    <key>FE00</key>
    <dict>
    <key>2018</key>
    <string>quoteleft</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>2018</key>
    <string>quoteleft.fwid</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>2019</key>
    <string>quoteright</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>2019</key>
    <string>quoteright.fwid</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>201C</key>
    <string>quotedblleft</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>201C</key>
    <string>quotedblleft.fwid</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>201D</key>
    <string>quotedblright</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>201D</key>
    <string>quotedblright.fwid</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>3001</key>
    <string>uni3001.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>3001</key>
    <string>uni3001</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>3002</key>
    <string>uni3002.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>3002</key>
    <string>uni3002</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF01</key>
    <string>uniFF01.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF01</key>
    <string>uniFF01</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF0C</key>
    <string>uniFF0C.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF0C</key>
    <string>uniFF0C</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF0E</key>
    <string>uniFF0E.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF0E</key>
    <string>uniFF0E</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF1A</key>
    <string>uniFF1A.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF1A</key>
    <string>uniFF1A</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF1B</key>
    <string>uniFF1B.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF1B</key>
    <string>uniFF1B</string>
    </dict>
    <key>FE00</key>
    <dict>
    <key>FF1F</key>
    <string>uniFF1F.crnr</string>
    </dict>
    <key>FE01</key>
    <dict>
    <key>FF1F</key>
    <string>uniFF1F</string>
    </dict>
  </dict>
```

</details>

aaronbell commented:
Thanks for the investigation. I think both should be modifiable via the post-fix python script. Will look into it.

aaronbell commented:
I've updated the export script to override the existing unicodeVariationSequences and to add the `meta` table with data produced by the latest FontCreator (15.0.3015). I don't love hardcoding them, but it is much simpler to do so, and figured they are unlikely to change in the future (or if they did, maybe the version of font creator used would also be updated at the same time).  

BTW, it appears that the `meta` table data was not being produced at all previously as it was not present in the source `.fcp` file. 

See what you think!

NightFurySL2001 commented:
> BTW, it appears that the meta table data was not being produced at all previously as it was not present in the source .fcp file.

That was correct as mentioned, but I am adding it to .fcp file on my end, which with FC 15.0.0.3004 caused `fontmake` to fail building due to incorrect format. The lib.plist file need to be edited to ensure `fontmake` can build correctly. I am fine with the hard coded part but I am changing the glyph names on my side too.

aaronbell commented:
Sounds good! If there’s any other changes you want to make, go for it, and then I think we can consider this one done.

NightFurySL2001 commented:
I think for the meta dlng and slng, this shouldn't be hardcoded in post, but instead be fixed before building from UFO since `fontmake` won't accept the UFO. Unless you prefer not putting the dlng and slng info in the .fcp file?

aaronbell commented:
I kind of feel that since the version of fcp can’t generate them correctly anyway it is easier to just override them (since the table is pretty simple) and if at some point we “upgrade” to a newer version, the override can be removed.

But adding the values into the source too is reasonable.

NightFurySL2001 commented:
Oh, I misread it and thought it was in post processing. My bad. I just need the updated GitHub Actions (or ill do it later) then should be good. I do have a local version pending to be merged though, so might take a while; Cactus Serif .fcp is overhauled too as it had various incorrect IVS, and will need an update to use the build system here too.

aaronbell commented:
Sounds good. Cactus isn’t a high priority for me at the moment but if it is easy enough to port the build system, I’ll do so.

NightFurySL2001 commented:
Can you help to check if this FCP file will match the requirements?
[ChocolateClassicalSans.zip](https://github.com/user-attachments/files/19872312/ChocolateClassicalSans.zip)

aaronbell commented:
I can check in the morning. If you've changed the UVS values, I'll need to re-generate those for my override script. Are there any other changes I should be aware of?

aaronbell commented:
I've been trying to get it building but it appears that there's issues with the OT features. First, all "fl" pairs are replaced with the `fl` ligature. And there's also a number of glyphs in the OT features that are not present in the font:
```
 Acircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:24)
 Ecircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:321)
 Icircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:595)
 Ocircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:765)
 Ucircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1081)
 acircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1480)
 ecircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1777)
 icircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2051)
 ocircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2221)
 ucircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2537)
 Acircumfiex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:48:297)
 fiorin (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:141:10)
```

Can you take a look at these?

aaronbell commented:
Sorry, I realized that I introduced some of the missing glyph bugs. 

To recap, I found that in the UFO generated by the older version of FCP, the /f/l pairs are being substituted with the `fl` ligature. It isn't clear to me why this is occurring as when I look at the feature source code in the newer version of FCP, they appear to be two separate glyphs. 

Anyway, I replaced them in the UFO, and proceeded with the build. 

It appears that there's other glyphs that are mentioned in the feature file but missing in the glyph set:
```
 Acircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:24)
 Ecircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:321)
 Icircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:595)
 Ocircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:765)
 Ucircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1081)
 acircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1480)
 ecircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:1777)
 icircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2051)
 ocircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2221)
 ucircumflex (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:46:2537)
 florin (first found at temp/ChocolateClassicalSans-Regular.ufo/features.fea:141:10)
```

If you could take a look. Thanks!

NightFurySL2001 commented:
![image](https://github.com/user-attachments/assets/2ba8e394-fd54-41f2-a8ca-726b74877682)
![image](https://github.com/user-attachments/assets/32f47cb9-7b19-49e9-8e7f-e34ca8d4c43f)
![image](https://github.com/user-attachments/assets/660bb6c8-72c5-4555-91bc-6a2c872c364d)

The listed glyphs are in the font as labeled here. The features exported from FC are validated and could be exported normally in FC with no missing glyphs.

MoonlitOwen commented:
Understand. Thanks a lot.

NightFurySL2001 commented:
Add permission bug fix and done
